### PR TITLE
fix: remove guava

### PR DESCRIPTION
**Pull request by effects-ai**
This PR removes `guava` from the list of dependencies.

### CODE REVIEW

File: NOTICE [hunk 1]
```diff
@@ -1 +1,32 @@
...
+Software Foundation distribution makes it eligible for export under the License Exception ENC Technology
+Software Unrestricted (TSU) exception (see the BIS Export Administration Regulations, Section 740.13) for
+both object code and source code.
+
+The following provides more details on the included cryptographic software:
+
+This software uses Bouncy Castle (https://www.bouncycastle.org/) for digital signatures and validation,
+certificates and certificate validation, key generation and management, and secure random number generation;
+and Java Security API (https://docs.oracle.com/en/java/javase/11/security/java-security-overview1.html)
+for message digests.
```
Comment (kpudlik) [inline NOTICE:32]: I think it should be a part of a separate commit/PR. Since the title is "fix: remove guava" it should only remove `guava` dependency.

I would put it to separate commit/PR called `docs: add NOTICE` or sth.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala [hunk 2]
```diff
@@ -25,11 +24,19 @@ object hash {
...
+      md.digest()
+    }
+
+    def fromBytes(bytes: Array[Byte]): Hash = {
+      val sha256Bytes = sha256FromBytes(bytes)
+      val sha256String = sha256Bytes
+        .foldLeft(new StringBuilder(64))((acc, x) => acc.append(s"%02x".format(x)))
+        .toString
+      Hash(sha256String)
+    }
```
Comment (kpudlik) [inline modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:39]: Have you verified if new methods produce same hashes as `Hashing` one?

**Review by kpudlik — CHANGES_REQUESTED** [review]

File: build.sbt [hunk 3]
```diff
@@ -20,7 +20,7 @@ val ghTokenSource = TokenSource.GitConfig("github.token") || TokenSource.Environ
 githubTokenSource := ghTokenSource
 
 lazy val commonSettings = Seq(
-  scalacOptions ++= List("-Ymacro-annotations", "-Yrangepos", "-Wconf:cat=unused:info", "-language:reflectiveCalls"),
+  scalacOptions ++= List("-Ymacro-annotations", "-Yrangepos", "-Wconf:cat=unused:info", "-language:reflectiveCalls", "-Wconf:msg=While parsing annotations in:silent"),
```
Comment (marcinwadon) [inline build.sbt:23]: Why do we need this? Does removing `guava` enforce adding this `Wconf` parameter?

File: modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala [hunk 4]
```diff
@@ -24,8 +24,8 @@ object address {
   object Address {
 
     def fromBytes(bytes: Array[Byte]): Address = {
-      val hashCode = Hash.hashCodeFromBytes(bytes)
-      val encoded = Base58.encode(hashCode.asBytes().toIndexedSeq)
+      val hashCode = Hash.sha256FromBytes(bytes)
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala:27]: Is it still a `hashCode`? You've changed the function name but not a variable

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:39]: I would be super super careful doing this. Hashing is the base of the whole codebase. We should test it again starting from genesis, revalidating the chain (partially at least) to make sure we won't make a breaking change and the hashing remains consistent.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala [hunk 5]
```diff
@@ -25,11 +24,19 @@ object hash {
...
-    def hashCodeFromBytes(bytes: Array[Byte]): HashCode =
-      Hashing.sha256().hashBytes(bytes)
-
-    def fromBytes(bytes: Array[Byte]): Hash =
-      Hash(hashCodeFromBytes(bytes).toString)
+    def sha256FromBytes(bytes: Array[Byte]): Array[Byte] = {
+      val md = MessageDigest.getInstance("SHA-256")
+      md.update(bytes)
+      md.digest()
+    }
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:31]: I think you should just change the implementation of the `hashCodeFromBytes`, not the fn name. We don't know who/how depends on that.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala [hunk 6]
```diff
@@ -25,11 +24,19 @@ object hash {
...
   object Hash {
 
-    def hashCodeFromBytes(bytes: Array[Byte]): HashCode =
-      Hashing.sha256().hashBytes(bytes)
-
-    def fromBytes(bytes: Array[Byte]): Hash =
-      Hash(hashCodeFromBytes(bytes).toString)
+    def sha256FromBytes(bytes: Array[Byte]): Array[Byte] = {
+      val md = MessageDigest.getInstance("SHA-256")
+      md.update(bytes)
```
Comment (marcinwadon) [inline modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:29]: Is it performant to create a new MessageDigest for each run? afaik google uses clone before each use, to use a single instance but keep it immutable/"clean" for external use.

can you run some benchmark to compare the old and new code? as I wrote in a separate comment, hashing is the most base level of the codebase and node runs it very very often.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 2 above)
Comment (effects-ai) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:39]: See https://github.com/effects-ai/tessellation/blob/sha256_tests/modules/shared/src/test/scala/io/constellationnetwork/security/GuavaJSATest.scala

File: modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala (hunk 4 above)
Comment (effects-ai) [reply modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala:27]: It's not an instance of `HashCode`, it's a byte array that contains a SHA-256 digest.

File: build.sbt (hunk 3 above)
Comment (effects-ai) [reply build.sbt:23]: It's needed to avoid this error:
```
[error] While parsing annotations in /Users/effects-ai/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/micrometer/micrometer-commons/1.10.2/micrometer-commons-1.10.2.jar(io/micrometer/common/lang/Nullable.class), could not find MAYBE in enum <none>.
[error] This is likely due to an implementation restriction: an annotation argument cannot refer to a member of the annotated class (scala/bug#7014).
[error] No warnings can be incurred under -Werror.
[error] two errors found
```

File: modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala (hunk 4 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/io/constellationnetwork/schema/address.scala:27]: Ok, so you should change the `hashCode` variable name too

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 6 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:29]: Ok, so what do we gain removing guava?

File: build.sbt (hunk 3 above)
Comment (marcinwadon) [reply build.sbt:23]: Ok, have you seen this error before removing guava? If so, it should be a separate change, not together with removing guava. Otherwise, it would be impossible to revert if needed. Also, it would mean it's your specific setup improperly configured, as everything is working correctly for everybody else over a long time.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:39]: What's the output?

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 5 above)
Comment (marcinwadon) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:31]: I agree, but you need to remember about compatibility too. We do have external dependencies on our code, so we cannot refactor everything as we like to.

File: build.sbt (hunk 3 above)
Comment (marcinwadon) [reply build.sbt:23]: btw:
```
[error] two errors found
```

What are those?

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 2 above)
Comment (effects-ai) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:39]: All green, since all tests use the same java.security.MessageDigest inside, plus

```
MessageDigest.clone() 32389 ms
MessageDigest.getInstance() 33180 ms
Guava 32336 ms
```

YMMV, though.

File: build.sbt (hunk 3 above)
Comment (effects-ai) [reply build.sbt:23]: > ```
> [error] While parsing annotations in /Users/effects-ai/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/micrometer/micrometer-commons/1.10.2/micrometer-commons-1.10.2.jar(io/micrometer/common/lang/Nullable.class), could not find MAYBE in enum <none>.
> [error] This is likely due to an implementation restriction: an annotation argument cannot refer to a member of the annotated class (scala/bug#7014).
> [error] No warnings can be incurred under -Werror.
> ```

Comment (effects-ai) [reply build.sbt:23]: > Ok, have you seen this error before removing guava? If so, it should be a separate change, not together with removing guava. Otherwise, it would be impossible to revert if needed. Also, it would mean it's your specific setup improperly configured, as everything is working correctly for everybody else over a long time.

I'll try to setup a completely different setup to verify.

File: modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala (hunk 5 above)
Comment (effects-ai) [reply modules/shared/src/main/scala/io/constellationnetwork/security/Hash.scala:31]: Such changes can be added to the release notes.

File: build.sbt (hunk 3 above)
Comment (effects-ai) [reply build.sbt:23]: It reproducible on a clean install of Ubuntu 22.0 as well.

This is another reason to drop guava - it hides issues with other dependencies. This issue affects all versions of `micrometer` up to 1.12, micrometer 1.13 seems to be fixed, but it requires code changes in `Metrics.scala`.

**Review by AlexBrandes — CHANGES_REQUESTED** [review]
Let's discuss at the meeting tomorrow but on the surface this change appears much too risky with a very small potential benefit. It would need extensive testing by the team to prove that it's valid and doesn't cause issues downstream. I don't think that it makes sense at this early stage in onboarding to the team.

**Comment by effects-ai** [comment]
Closing this PR in favor of https://github.com/Constellation-Labs/tessellation/pull/965
